### ISSUE TITLE
Fully document support for test interfaces

### ISSUE THREAD

**Issue by sbrannen** [junit-team/junit5#281]
## Status Quo

The User Guide currently contains a section on [Interface Default Methods](http://junit.org/junit5/docs/current/user-guide/#interface-default-methods); however, it does not mention all features supported in _test interfaces_.
## Deliverables
- [x] Document support for configuring extensions and tags in test interfaces via `@ExtendWith` and `@Tag`.
- [x] Document support for declaring `@BeforeAll`, `@AfterAll`, `@BeforeEach`, and `@AfterEach` methods in test interfaces.
- [x] Document support for declaring `@TestFactory` methods in test interfaces.

**Comment by nicolaiparlog** [comment]
Is this feature really worth the effort? It is very easy to do the same with nested classes (one abstract test class per interface; implementations are nested) or dynamic tests (test factory per interface) but without the added complexity (colliding default methods, no `@...All`).

**Comment by sbrannen** [comment]
The feature already exists.

This issue is merely about documenting existing support for test interfaces.

If you don't like them, you don't have to use them.

But they certainly have one big advantage over abstract base classes and dynamic tests, namely _multiple inheritance_. Furthermore, you can extend any class like you, which is in direct contrast to using abstract base test classes.

If that doesn't appeal to you, feel free to use abstract base classes, `@Nested` classes, dynamic tests, or whatever suits your needs.

In the end, it's all about choice. 😉

**Comment by smoyer64** [comment]
I'm actually looking forward to this feature ... with the
multiple-inheritance Sam mentioned above (using the default methods allowed
by Java 8), you can effectively treat test interfaces like the traits that
are common to testing with the Scala framework.

On Thu, May 26, 2016 at 9:06 AM, Sam Brannen <notifications@github.com>
wrote:

> The feature already exists.
>
> This issue is merely about documenting existing support for test
> interfaces.
>
> If you don't like them, you don't have to use them.
>
> But they certainly have one big advantage over abstract base classes and
> dynamic tests, namely *multiple inheritance*. Furthermore, you can extend
> any class like you, which is in direct contrast to using abstract base test
> classes.
>
> If that doesn't appeal to you, feel free to use abstract base classes,
> @Nested classes, dynamic tests, or whatever suits your needs.
>
> In the end, it's all about choice. 😉
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/junit-team/junit5/issues/281#issuecomment-221865036>
>



-- 
Sent from my iBerry MacPhone

**Comment by nicolaiparlog** [comment]
It's a nice feature, don't get me wrong. I just don't think that it adds enough value. Yes, multiple inheritance is nice (unless methods collide). But you can easily do something similar with nested classes:

```
class MyClassTest {
    @Nested
    class MyClassEqualsContract extends EqualsContract { ... }

    @Nested
    class MyClassComparatorContract extends ComparatorContract { ... }
}
```

I find this much more explicit and it makes clearer which abstract methods belong to which tests.

> In the end, it's all about choice.

This is _the_ argument for feature creep. Again, this is definitely a nice-to-have. But does it [prefer extension points over features](https://github.com/junit-team/junit5/wiki/Core-Principles#prefer-extension-points-over-features)? I am just making an argument for a leaner JUnit, feel free to dismiss it. :smile_cat:

**Comment by sbrannen** [comment]
_in progress_ due to #606

**Comment by sbrannen** [comment]
Implemented in f4b67dacc8dab2b2fa6499522f6f480c20fc1f31.
